snuba/query/matchers.py

```diff
@@ -0,0 +1,317 @@
+    and returns the first valid one.
+    """
+
+    patterns: Sequence[Pattern[TMatchedType]]
+
+    def match(self, node: AnyType) -> Optional[MatchResult]:
+        for p in self.patterns:
+            ret = p.match(node)
+            if ret:
+                return ret
+        return None
+
+
+@dataclass(frozen=True)
+class Column(Pattern[ColumnExpr]):
+    """
+    Matches a Column in an AST expression.
+    The column is defined by alias, name and table. For each, a Pattern can be
+    provided. If one of these Patterns is left None, that field will be ignored
+    when matching (equivalent to Any, but less verbose).
+    """
+
+    alias: Optional[Pattern[Optional[str]]] = None
+    column_name: Optional[Pattern[str]] = None
+    table_name: Optional[Pattern[Optional[str]]] = None
+
+    def match(self, node: AnyType) -> Optional[MatchResult]:
+        if not isinstance(node, ColumnExpr):
+            return None
+
+        result = MatchResult()
+        for pattern, value in (
+            (self.alias, node.alias),
+            (self.table_name, node.table_name),
+        ):
+            if pattern is not None:
+                partial_result = pattern.match(value)
+                if not partial_result:
+                    return None
+                result = result.merge(partial_result)
+
+        # column_name does not have the same type as alias and table_name
+        # so this case cannot be handled in the loop above as it does not
+        # type check.
+        if self.column_name is not None:
+            partial_result = self.column_name.match(node.column_name)
+            if not partial_result:
+                return None
+            result = result.merge(partial_result)
+
+        return result
+
+
+@dataclass(frozen=True)
+class Literal(Pattern[LiteralExpr]):
+    alias: Optional[Pattern[Optional[str]]] = None
+    value: Optional[Pattern[OptionalScalarType]] = None
+
+    def match(self, node: AnyType) -> Optional[MatchResult]:
+        if not isinstance(node, LiteralExpr):
+            return None
+
+        result = MatchResult()
+        for pattern, value in (
+            (self.alias, node.alias),
+            (self.value, node.value),
+        ):
+            if pattern is not None:
+                partial_result = pattern.match(value)
+                if not partial_result:
+                    return None
+                result = result.merge(partial_result)
+
+        return result
+
+
+@dataclass(frozen=True)
+class FunctionCall(Pattern[FunctionCallExpr]):
+    """
+    Matches a Function in the AST expression.
+    It works like the Column Pattern. if alias, function_name and function_parameters
+    are provided, they have to match, otherwise they are ignored.
+    """
+
+    alias: Optional[Pattern[Optional[str]]] = None
+    function_name: Optional[Pattern[str]] = None
+    parameters: Optional[Tuple[Pattern[Expression], ...]] = None
```

Comment:
I already forgot about this and had to go back to https://github.com/getsentry/snuba/pull/942#discussion_r427556300, it might make sense just to note this in comments for future reference. (I could see someone trying to use this later and having the same question.)